ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -348,6 +374,186 @@ protected Map<DataNodeId, List<RemoteReplicaInfo>> selectReplicas(Map<DataNodeId
+      }
+    }
+
+    if (!running || shouldTerminateCurrentCycle) {
+      return;
+    }
+
+    int maxIterationCount = 0;
+    for (Map.Entry<Integer, Integer> entry : groupIdIterationCountMap.entrySet()) {
+      maxIterationCount = Integer.max(maxIterationCount, entry.getValue());
+    }
+
+    if (maxIterationCount == maxIterationsPerGroupPerCycle) {
+
+      if (maxIterationsReached.isFalse()) {
+        maxIterationsReached.setTrue();
+        if (inflightRemoteReplicaGroup.isEmpty()) {
+          allReplicasCaughtUpEarly.setTrue();
+        }
+      }
+
+      return;
+    }
+
+    for (Map.Entry<Integer, List<RemoteReplicaInfo>> entry : groupIdToRemoteReplicaMap.entrySet()) {
+      Integer groupId = entry.getKey();
+      List<RemoteReplicaInfo> remoteReplicas = entry.getValue();
+      DataNodeId remoteNode = remoteReplicas.get(0).getReplicaId().getDataNodeId();
+
+      Integer standbyReplicasNoProgressReplicaGroupId = dataNodeIdStandbyReplicasNoProgressReplicaGroupId.get(remoteNode);
+
+      Set<RemoteReplicaInfo> nonIteratingNonPendingRemoteReplicas = new LinkedHashSet<>(remoteReplicas);
+
+      remoteReplicas.forEach(remoteReplicaInfo -> {
+        if(remoteReplicaToThrottledTill.containsKey(remoteReplicaInfo) && remoteReplicaToThrottledTill.get(remoteReplicaInfo) > time.milliseconds()){
+          nonIteratingNonPendingRemoteReplicas.remove(remoteReplicaInfo);
+          throttleCount.inc();
+        }
+      });
+
+      if(inflightRemoteReplicaGroup.containsKey(groupId)){
+        nonIteratingNonPendingRemoteReplicas.removeAll(inflightRemoteReplicaGroup.get(groupId).remoteReplicaInfos);
+      }
+
+      if(inflightRemoteReplicaGroup.containsKey(standbyReplicasNoProgressReplicaGroupId)) {
+        nonIteratingNonPendingRemoteReplicas.removeAll(inflightRemoteReplicaGroup.get(standbyReplicasNoProgressReplicaGroupId).remoteReplicaInfos);
+      }
+
+      List<RemoteReplicaInfo> activeReplicas = new ArrayList<>();
+      List<RemoteReplicaInfo> standbyReplicasWithNoProgress = new ArrayList<>();
+
+      filterRemoteReplicasToReplicate(new ArrayList<>(nonIteratingNonPendingRemoteReplicas), activeReplicas, standbyReplicasWithNoProgress);
+
+      if (!inflightRemoteReplicaGroup.containsKey(groupId)) {
+        if (!activeReplicas.isEmpty()) {
+          RemoteReplicaGroup groupForActiveReplicas = new RemoteReplicaGroup(activeReplicas, remoteNode, false, groupId);
+          inflightRemoteReplicaGroup.put(groupId, groupForActiveReplicas);
+          groupIdIterationCountMap.put(groupId, groupIdIterationCountMap.getOrDefault(groupId, 0) + 1);
+        }
+      }
+
+      List<RemoteReplicaInfo> standByReplicasTimedOutOnNoProgress =
+          getRemoteStandbyReplicasTimedOutOnNoProgress(standbyReplicasWithNoProgress);
+
+      if (standByReplicasTimedOutOnNoProgress.isEmpty()) {
+        continue;
+      }
+
+      dataNodeIdToPendingReplicasMarkedForStandByNoProgress.putIfAbsent(remoteNode, new HashSet<>());
```

Comment:
consider smaller fn names, this is just hard to read

Reply:
changed the name.